Docs: Add basic Gateway API troubleshooting (#25945)
cilium-team/cilium: Documentation/network/servicemesh/gateway-api/troubleshooting.rst

> These steps are for troubleshooting issues with Gateway API resources. 
> It is important to also have followed the Generic and Setup Verification steps from the :ref:`Troubleshooting Ingress & Service Mesh page <troubleshooting_servicemesh>`  
> This part of the documentation will guide you through the different mechanics of Gateway API and how to troubleshoot them.

Reviewer: Use clear, plain language
```suggestion
This page guides you through the different mechanics of Gateway API and how to troubleshoot them.

Be sure to follow the Generic and Setup Verification steps from the :ref:`Troubleshooting Ingress & Service Mesh page<troubleshooting_servicemesh>`.
```

> Checking the resources

Reviewer: ```suggestion
Checking resources
```

----------------------

#. Check the Gateway resource 

    .. code-block:: shell-session

      $ kubectl get gateway -A
      NAMESPACE                   NAME                 CLASS    ADDRESS          PROGRAMMED   AGE
      website                     http-gateway         cilium   172.21.255.202   True         5h
      webshop                     tls-gateway          cilium   172.21.255.203   True         5h
  

>     This will get you an overview for all the Gateways in the cluster. There are a few important tings to check:

Reviewer: Specific is better than general, use present tense, typo
```suggestion
    The preceding command returns an overview of all the Gateways in the cluster. Check the following:
```

>     * Is the Gateway programmed? This means that Cilium saw the gateway and prepared configuration for it.
>       
>       * If this is missing make sure that Gateway API is enabled in the Cilium configuration.

Reviewer: You are overusing "this" and it isn't clear what you mean each time. For example:
> If this is missing

If what is missing, specifically?
```suggestion
    * Is the Gateway programmed? 
    
        A programmed Gateway means that Cilium saw the gateway and prepared a configuration for it.
      
      * If this is missing make sure that Gateway API is enabled in the Cilium configuration.
```

This page guides you through the different mechanics of Gateway API and how to troubleshoot them.

Be sure to follow the Generic and Setup Verification steps from the :ref:`Troubleshooting Ingress & Service Mesh page<troubleshooting_servicemesh>`.

Checking resources
----------------------

#. Check the Gateway resource 

    .. code-block:: shell-session

      $ kubectl get gateway -A
      NAMESPACE                   NAME                 CLASS    ADDRESS          PROGRAMMED   AGE
      website                     http-gateway         cilium   172.21.255.202   True         5h
      webshop                     tls-gateway          cilium   172.21.255.203   True         5h
  
    The preceding command returns an overview of all the Gateways in the cluster. Check the following:

    * Is the Gateway programmed? 
    

>         A programmed Gateway means that Cilium saw the gateway and prepared a configuration for it.

Reviewer: clarity
```suggestion
        A programmed Gateway means that Cilium prepared a configuration for it.
```

>       * If the ``Programmed true`` indicator is missing make sure that Gateway API is enabled in the Cilium configuration.

Reviewer: comma nit
```suggestion
      * If the ``Programmed true`` indicator is missing, make sure that Gateway API is enabled in the Cilium configuration.
```

  
    * Does the gateway have an address? If you see one it means that that a LoadBalancer service got assigned to the gateway.
  

>       * You can check the service with ``kubectl get service``, if there is no IP there either you might be missing a LoadBalancer implementation.

Reviewer: Formatting, clarity, present tense
```suggestion
    * Does the gateway have an address?
    
    You can check the service with ``kubectl get service``. 
    If the gateway has an address, it means that a LoadBalancer service is assigned to the gateway.
    If no IP appears, you might be missing a LoadBalancer implementation.
  

```

  
    * Is the class ``cilium``? Cilium will only program Gateways with the class ``cilium``.

>     * In case the resource type is not found, make sure that the Gateway API CRDs are installed.

Reviewer: Formatting and clarity, present tense

Resource type for what? Be specific.
```suggestion
    * Is the class ``cilium``? 
    
    Cilium only program Gateways with the class ``cilium``.
    
    * If the resource type is not found, make sure that the Gateway API CRDs are installed.
```

>     If you need more info to further investigate the issue you can always use ``kubectl describe gateway``

Reviewer: Clarity
```suggestion
    You can use ``kubectl describe gateway`` to investigate issues more thoroughly.
```

    .. code-block:: shell-session

      $ kubectl describe gateway <name>

        Conditions:
          Message:               Gateway successfully scheduled
          Reason:                Accepted
          Status:                True
          Type:                  Accepted
          Message:               Gateway successfully configured
          Reason:                Programmed
          Status:                True
          Type:                  Programmed
          [...]
        Listeners:
          Attached Routes:  2
          Conditions:
            Message:               Listener Ready
            Reason:                Programmed
            Status:                True
            Type:                  Programmed
            Message:               Listener Accepted
            Reason:                Accepted
            Status:                True
            [...]

    You'll see under the status the general status of the Gateway as well as the status of the configured listeners.
    Per listener you also get a count of how many routes are successfully attached to it.

>     Per Gateway or listener you can always see two status conditions:

Reviewer: Clear, plain language, present tense
```suggestion
    You can see the general status of the gateway as well as the status of the configured listeners.
    
    Listener status displays the number of routes successfully attached to the listener.
    
    You can see status conditions for both gateway and listener:
```

      * ``Accepted``: this means that the Gateway configuration was accepted.
      * ``Programmed``: this means that the Gateway configuration was programmed into Envoy.
      * ``ResolvedRefs``: this means that all referenced secrets were found and got permission to be used.
  

>     If any of these are set to false you can get more information by looking at the ``Message`` and ``Reason`` fields.

Reviewer: Clarity
```suggestion
      * ``Accepted``: the Gateway configuration was accepted.
      * ``Programmed``: the Gateway configuration was programmed into Envoy.
      * ``ResolvedRefs``: all referenced secrets were found and have permission for use.
  
    If any of these conditions are set to false, the ``Message`` and ``Reason`` fields give more information.
```

#. Check the HTTPRoute resource

  When our Gateway is functional we can check our routes to see if they are configured correctly.

>   The way to do check the route status is very similar to the Gateway resource. While these instructions are written for HTTPRoute, they also apply to other route types.

Reviewer: Address the user as "you"
```suggestion
  When the Gateway is functional, you can check the routes to verify if they are configured correctly.
  The way to check route status is similar to checking the status of a gateway resource. 
  
  While these instructions are written for HTTPRoute, they also apply to other route types.
```

  .. code-block:: shell-session

    $ kubectl get httproute -A
    NAMESPACE                 NAME              HOSTNAMES         AGE
    website                   homepage          www.example.org   17m
    webshop                   catalog-service                     17m
    webshop                   cart-service                        17m

>   To get further information we can again use ``kubectl describe httproute <name>``

Reviewer: Address the user as "you", period nit
```suggestion
  To get more information, enter ``kubectl describe httproute <name>``.
```

  .. code-block:: shell-session

    $ kubectl describe httproute <name>
    Status:
      Parents:
        Conditions:
          Last Transition Time:  2023-06-05T15:11:53Z
          Message:               Accepted HTTPRoute
          Observed Generation:   1
          Reason:                Accepted
          Status:                True
          Type:                  Accepted
          Last Transition Time:  2023-06-05T15:11:53Z
          Message:               Service reference is valid
          Observed Generation:   1
          Reason:                ResolvedRefs
          Status:                True
          Type:                  ResolvedRefs
        Controller Name:         io.cilium/gateway-controller
        Parent Ref:
          Group:  gateway.networking.k8s.io
          Kind:   Gateway
          Name:   same-namespace

>   Here under status we get all the conditions that are relevant for the specific HTTPRoute.
>   These are listed per parent reference to the gateway. If you linked the route to multiple gateways you will see multiple entries here.
>   There are two status conditions that are of interest to investigate:

Reviewer: Clarity, address the user as "you"

There are multiple fields with "Accepted" and "ResolvedRefs". Which ones should users pay attention to?
```suggestion
  Status lists the conditions that are relevant for the specific HTTPRoute.
  Conditions are listed by parent reference to the gateway. If you linked the route to multiple gateways, multiple entries appear.
  Notice the following conditions:
```

>   * ``Accepted``: This means that the HTTPRoute configuration was correct and accepted.
>   * ``ResolvedRefs``: This means that the referenced services were found and are valid references.

Reviewer: clarity
```suggestion
  * ``Accepted``: The HTTPRoute configuration was correct and accepted.
  * ``ResolvedRefs``: The referenced services were found and are valid references.
```

  
  If any of these are set to false you can get more information by looking at the ``Message`` and ``Reason`` fields.

Common mistakes
---------------

.. include:: mistakes-warning.rst

> * The backend service does not exist. You can check if the backend service was found under the ``conditions`` field when doing a ``kubectl describe httproute <name>``

Reviewer: formatting, period nit, active voice

What's doing the finding? Is it Kubernetes via kubectl? It's better to be as specific as possible.
```suggestion
* The backend service does not exist. 

    To verify whether the backend service was found, run ``kubectl describe httproute <name>`` and inspect the ``conditions`` field:
```

  
  .. code-block:: shell-session

        Parents:
          Conditions:
            Last Transition Time:  2023-06-06T13:55:10Z
            Message:               Service "backend" not found
            Observed Generation:   1
            Reason:                BackendNotFound
            Status:                False
            Type:                  ResolvedRefs
            Last Transition Time:  2023-06-06T13:55:10Z
            Message:               Accepted HTTPRoute
            Observed Generation:   1
            Reason:                Accepted
            Status:                True
            Type:                  Accepted
          Controller Name:         io.cilium/gateway-controller

> * The Gateway specified under ``parentRefs`` does not exist. You check if the gateway was found under the ``conditions`` field when doing a ``kubectl describe httproute <name>``

Reviewer: Formatting nit

Similar to the previous review item ☝🏻: what's doing the finding of the gateway? Kubernetes via kubectl? It's better to be as specific as possible.
```suggestion
* The gateway specified under ``parentRefs`` does not exist.

    To verify whether the gateway was found, run ``kubectl describe httproute <name>`` and inspect the ``conditions`` field:
```

   
.. code-block:: shell-session

  Parents:
    Conditions:
      Last Transition Time:  2023-06-06T13:56:40Z
      Message:               Gateway.gateway.networking.k8s.io "my-gatewai" not found
      Observed Generation:   2
        Reason:                InvalidHTTPRoute
        Status:                False
        Type:                  Accepted

Underlying mechanics: a high level overview
-------------------------------------------

> The deployment of Cilium has two parts that handle Gateway API resources: the Cilium agent and the Cilium operator.
> The Cilium operator will watch all Gateway API resources. After which it will verify if the resources are valid.
> If they are valid they will get marked as accepted. That starts the process of translation into Cilium Envoy Configuration resources.
> 
> The Cilium Envoy Configuration resources are then picked up by the Cilium agent.
> The Cilium agent will use these to supply the built in Envoy or the Envoy DaemonSet with the configuration. Envoy is where traffic will be handled.

Reviewer: Clarity, present tense
```suggestion
A Cilium deployment has two parts that handle Gateway API resources: the Cilium agent and the Cilium operator.

The Cilium operator watches all Gateway API resources and verifies whether the resources are valid.
If resources are valid, the operator marks them as accepted. That starts the process of translation into Cilium Envoy Configuration resources.

The Cilium agent then picks up the Cilium Envoy Configuration resources.

The Cilium agent uses the resources to supply the configuration to the built in Envoy or the Envoy DaemonSet. Envoy handles traffic.
```

Checking resources
----------------------

#. Check the Gateway resource 

    .. code-block:: shell-session

      $ kubectl get gateway -A
      NAMESPACE                   NAME                 CLASS    ADDRESS          PROGRAMMED   AGE
      website                     http-gateway         cilium   172.21.255.202   True         5h
      webshop                     tls-gateway          cilium   172.21.255.203   True         5h
  
    The preceding command returns an overview of all the Gateways in the cluster. Check the following:

    * Is the Gateway programmed? 
    
        A programmed Gateway means that Cilium prepared a configuration for it.
      
      * If the ``Programmed true`` indicator is missing, make sure that Gateway API is enabled in the Cilium configuration.
  
    * Does the gateway have an address?
    
    You can check the service with ``kubectl get service``. 
    If the gateway has an address, it means that a LoadBalancer service is assigned to the gateway.
    If no IP appears, you might be missing a LoadBalancer implementation.
  

> 

Reviewer: Remove new lines

  
    * Is the class ``cilium``? 
    

>     Cilium only program Gateways with the class ``cilium``.

Reviewer: s/program/programs

Reviewer: I've opened a separate issue to talk about capitalization of API objects, in #26904.

          Reason:                Programmed
          Status:                True
          Type:                  Programmed
          [...]
        Listeners:
          Attached Routes:  2
          Conditions:
            Message:               Listener Ready
            Reason:                Programmed
            Status:                True
            Type:                  Programmed
            Message:               Listener Accepted
            Reason:                Accepted
            Status:                True
            [...]

    You can see the general status of the gateway as well as the status of the configured listeners.
    
    Listener status displays the number of routes successfully attached to the listener.
    
    You can see status conditions for both gateway and listener:

      * ``Accepted``: the Gateway configuration was accepted.
      * ``Programmed``: the Gateway configuration was programmed into Envoy.
      * ``ResolvedRefs``: all referenced secrets were found and have permission for use.
  
    If any of these conditions are set to false, the ``Message`` and ``Reason`` fields give more information.

> 

Reviewer: Remove newline
```suggestion
```

  To get more information, enter ``kubectl describe httproute <name>``.

  .. code-block:: shell-session

    $ kubectl describe httproute <name>
    Status:
      Parents:
        Conditions:
          Last Transition Time:  2023-06-05T15:11:53Z
          Message:               Accepted HTTPRoute
          Observed Generation:   1
          Reason:                Accepted
          Status:                True
          Type:                  Accepted
          Last Transition Time:  2023-06-05T15:11:53Z
          Message:               Service reference is valid
          Observed Generation:   1
          Reason:                ResolvedRefs
          Status:                True
          Type:                  ResolvedRefs
        Controller Name:         io.cilium/gateway-controller
        Parent Ref:
          Group:  gateway.networking.k8s.io
          Kind:   Gateway
          Name:   same-namespace

  Status lists the conditions that are relevant for the specific ``HTTPRoute``.
  Conditions are listed by parent reference to the gateway. If you linked the route to multiple gateways, multiple entries appear.

>   You will see a group of ``Reason``, ``Type``, ``Status`` and ``Message``. Where ``Type`` will indicate the condition type and ``Status``
>   will tell you it is met in a boolean. Optionally, ``Message`` will give you more information about the condition.

Reviewer: Present tense, new line
```suggestion
  Conditions include ``Reason``, ``Type``, ``Status`` and ``Message``. ``Type`` indicates the condition type, and ``Status`` indicates with a boolean whether the condition type is met. Optionally, ``Message`` gives you more information about the condition.
  
```

  Notice the following condition types:

  * ``Accepted``: The HTTPRoute configuration was correct and accepted.
  * ``ResolvedRefs``: The referenced services were found and are valid references.
  

>   If any of these are set to false you can get more information by looking at the ``Message`` and ``Reason`` fields.

Reviewer: comma nit
```suggestion
  If any of these are set to false, you can get more information by looking at the ``Message`` and ``Reason`` fields.
```

Common mistakes
---------------

> 

Reviewer: Remove newline
```suggestion
```
